### PR TITLE
[MODEL] Add Falcon H1

### PR DESCRIPTION
**Pull request by younesbelkada**

### CODE REVIEW

File: src/transformers/models/falcon_h1/configuration_falcon_h1.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        mamba_rms_norm (`<fill_type>`, *optional*, defaults to `False`): <fill_docstring>
+        projectors_bias (`<fill_type>`, *optional*, defaults to `False`): <fill_docstring>
+        rope_theta (`<fill_type>`, *optional*, defaults to 100000.0): <fill_docstring>
+        rope_scaling (`<fill_type>`, *optional*): <fill_docstring>
+        lm_head_multiplier (`<fill_type>`, *optional*, defaults to 1.0): <fill_docstring>
+        embedding_multiplier (`<fill_type>`, *optional*, defaults to 1.0): <fill_docstring>
+        mlp_multipliers (`<fill_type>`, *optional*): <fill_docstring>
+        key_multiplier (`<fill_type>`, *optional*): <fill_docstring>
+        attention_out_multiplier (`<fill_type>`, *optional*): <fill_docstring>
+        attention_in_multiplier (`<fill_type>`, *optional*): <fill_docstring>
```
Comment (ArthurZucker) [inline src/transformers/models/falcon_h1/configuration_falcon_h1.py:113]: to fill

File: tests/models/falcon_h1/test_modeling_falcon_h1.py [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            # They should result in very similar logits
+            torch.testing.assert_close(next_logits_wo_padding, next_logits_with_padding, rtol=1e-5, atol=1e-5)
+
+
+@slow
+@require_torch
+@require_torch_gpu
+class FalconH1ModelIntegrationTest(unittest.TestCase):
+    # TODO: add integration tests for all model sizes
+    pass
```
Comment (ArthurZucker) [inline tests/models/falcon_h1/test_modeling_falcon_h1.py]: missing before we can merge

**Review by ArthurZucker — COMMENTED** [review]
LGTM

File: src/transformers/models/falcon_h1/configuration_falcon_h1.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/models/falcon_h1/configuration_falcon_h1.py]: Done

File: tests/models/falcon_h1/test_modeling_falcon_h1.py (hunk 2 above)
Comment (dhiaEddineRhaiem) [reply tests/models/falcon_h1/test_modeling_falcon_h1.py]: done @ArthurZucker !

**Comment by ArthurZucker** [comment]
Thanks all for the contribution!!! 🤗

**Comment by ArthurZucker** [comment]
@DarkLight1337 we just did one yesterday 😢

**Comment by ArthurZucker** [comment]
We are gonna do a model based release!

**Comment by ydshieh** [comment]
@younesbelkada `test_llama_3_1_hard` 😆

**Comment by ydshieh** [comment]
I am getting the following output, is it normal?

```
(Pdb) print(generated_text)
user
Tell me about the french revolution.
     of U PT  SI      di  the  the Sl    D  �  " ACK \   S  S   S     MOD OF A D\  Ont reli

 \   \   S$
 \) S   es\ $ $\  S\ S\) �  (?   S shortens \ {        $ " $$$ $$                            }{ $       $$$ $] $   $]]$\ � $\   ES    � ] ]  Sus  Space  of Corpor $  Of  Un    User   User Item Un
    Box Med   Med      Med Right or  )                      splits     halfway                                    Box  ]  �   �    ∩                stack               Y ma Bu    for  )  1 User Item—     U  ] S      Food      for     $ �           for       S     O

```

**Comment by dhiaEddineRhaiem** [comment]
Thank you for pointing out the typo, @ydshieh.
It has now been addressed and fixed in this [PR](https://github.com/huggingface/transformers/pull/38256)

**Comment by ydshieh** [comment]
@dhiaEddineRhaiem Do you want the test or you check it with a different code snippet? I am getting the the same strange outputs both with T4 and A10.

**Comment by dhiaEddineRhaiem** [comment]
hello again@ydshieh 
thanks for your remark!

one other potential cause is that **Falcon H1 is particularly sensitive to temperature changes above 0.3 or 0.4**, likely because it already produces **well-calibrated and sharply peaked logits by default**, Basically:

🔹 Its raw logits are already **well-separated**, so lowering temperature (e.g. to 0.1) **keeps that separation strong** → **stable behavior**.

🔹 Increasing T > 0.3 or 0.4 **flattens that**, letting weaker tokens sneak in → **instability**.

Ampirically, i would advise to set **T=0.1** !

**Comment by ydshieh** [comment]
oh my god .... that is going be tough this issue. Thank you for checking.

Could you first share your machine type? T4/A10/A100/H100 etc?

And copy paste the output of `transformers-cli env` 🙏 ?

**Comment by dhiaEddineRhaiem** [comment]
1. Machine is `H100`
2. `transformers-cli env` gives :
```
- `transformers` version: 4.53.0.dev0
- Platform: Linux-6.6.72+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.31.4
- Safetensors version: 0.4.5
- Accelerate version: 1.1.1
- Accelerate config:    not found
- DeepSpeed version: not installed
- PyTorch version (GPU?): 2.4.0a0+3bcc3cddb5.nv24.07 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA H100 80GB HBM3
```

**Comment by ydshieh** [comment]
I am seeing

> The fast path is not available because on of `(selective_state_update, causal_conv1d_fn, causal_conv1d_update)` is None. Falling back to the naive implementation. To install follow https://github.com/state-spaces/mamba/#installation and https://github.com/Dao-AILab/causal-conv1d

maybe it's the cause. I will check further

**Comment by dhiaEddineRhaiem** [comment]
i think also it is the cause , we saw similar behaviour locally when fast path is not used.
please try to install mamba-ssm and causal-conv1d and retry.

**Comment by ydshieh** [comment]
Sure, but a question: do we want to maintain the slow path (even if not identical results, at least outputs that look normal ..?)

I change the prompts, and still get non-sense outputs 😰

**Comment by dhiaEddineRhaiem** [comment]
i would say we might want to retain the **slow path**, as the same logic applies across **hybrid and pure Mamba2 models.**


Mamba2 is highly sensitive to **numerical precision**: key components like A**_log, dt, and internal ssm_states** operate in **fp32**. Without the Triton fast path, the fallback accumulates precision errors across tokens — especially in long contexts — which likely may lead the degraded outputs. 

we will further debug internally to see if other causes may be found and overcomed.

**Comment by ydshieh** [comment]
Got it. I confirmed that with fast path, the outputs look normal.

**Comment by vasqu** [comment]
@dhiaEddineRhaiem I think the slow path is missing a recent fix, see #37533 

Tl;dr: the repeat pattern on the mamba heads has been wrong --> we need a repeat interleave, not a simple repeat

Will try to make the mamba model paths inheritable in the future. The copy pasting atm is very error prone atm 😢

**Comment by dhiaEddineRhaiem** [comment]
Hey @vasqu,

 Many Thanks for pointing this out:
1. i confirm torch.repeat gives undesirable repeat pattern in the mamba heads
2. Another major issue is that mup multipliers that we released with the weights are not included in the slow path fwd pass which explains the wierd generations.

 I will soon raise a PR to fix that.
 Thanks again.

**Comment by ArthurZucker** [comment]
thanks all for diving! 🚀

**Comment by dhiaEddineRhaiem** [comment]
hello @vasqu , @ydshieh , 
as discussed here is the [PR](https://github.com/huggingface/transformers/pull/38320) fixing the two problems i found in the slow path of FalconH1
